### ISSUE TITLE
Zelda Ocarina of Time crashing when the deku tree is talking.

### ISSUE THREAD

**Issue by Ncapron** [citra-emu/citra#2447]
When you leave the deku tree after beating Ghoma you got the heart, that gives you a 4th heart for your HP. And then you go back outside the tree and he says something like "do you want me to tell you more?" you have the choice between yes or no. and Then the game crash with this on the console. 

Btw : I'm playing on mac with Nightly Build - 55c91ae version

''
> ped buffers are unimplemented at the moment
> [ 125.866441] Audio.DSP <Error> audio_core/hle/source.cpp:DequeueBuffer:259: Looped buffers are unimplemented at the moment
> [ 126.094656] Audio.DSP <Error> audio_core/hle/source.cpp:DequeueBuffer:259: Looped buffers are unimplemented at the moment
> [ 126.628569] Audio.DSP <Error> audio_core/hle/source.cpp:DequeueBuffer:259: Looped buffers are unimplemented at the moment
> [ 126.700078] Audio.DSP <Error> audio_core/hle/source.cpp:DequeueBuffer:259: Looped buffers are unimplemented at the moment
> [ 126.894703] Audio.DSP <Error> audio_core/hle/source.cpp:DequeueBuffer:259: Looped buffers are unimplemented at the moment
> [ 127.785783] Debug <Critical> video_core/renderer_opengl/gl_rasterizer.cpp:operator():1084: Assertion Failed!
> Uniform block size did not match! Got 0, expected 960
> Trace/BPT trap: 5
> logout
> Saving session...
> ...copying shared history...
> ...saving history...truncating history files...
> ...completed.
> Deleting expired sessions...none found.
> 
> [Opération terminée]
''

**Comment by prasoc** [comment]
Could you provide me with the saved session and detail what version/region of OoT you are using so I can reproduce this issue

Thanks,
James

**Comment by Ncapron** [comment]
what do you mean by saved session?

I'm playing with the 3DS OoT the Europe one with Nightly Build - 55c91ae version for citra.

Hope it will be usefull,

Nicolas

**Comment by RavenHome1** [comment]
i think he means the game save file (upload it here), also this issue was posted a lot on forums but it was gone with the forum updates

**Comment by prasoc** [comment]
yup, just to be able to quickly test and debug the issue. I can get the right ROM version, but a quick save that I can load will do wonders for fixing

**Comment by Ncapron** [comment]
you'll think i'm dumb :D but the save is with the .3DS or there is a folder somewhere just for the save fild

**Comment by prasoc** [comment]
Well the SD card is stored here:
C:\Users\{{user}}\AppData\Roaming\Citra\sdmc

The actual OoT savegames are stored here (for me, atleast):
C:\Users\jb27g\AppData\Roaming\Citra\sdmc\Nintendo 3DS\00000000000000000000000000000000\00000000000000000000000000000000\title\00040000\00033500\data\00000001

Upload save00.bin and system.dat from that folder and I can replicate your issue

**Comment by legoj15** [comment]
You can access a game's save by right clicking it on the list in citra-qt, it will bring you to the folder where the save files are

**Comment by Ncapron** [comment]
https://www.dropbox.com/s/39s7hziampuqqel/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.15.png?dl=0

https://www.dropbox.com/s/h0zb65epl9721zf/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.21.png?dl=0

**Comment by Ncapron** [comment]
https://www.dropbox.com/s/39s7hziampuqqel/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.15.png?dl=0

https://www.dropbox.com/s/h0zb65epl9721zf/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.21.png?dl=0

https://www.dropbox.com/s/cklma7g673gjwly/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.28.png?dl=0

https://www.dropbox.com/s/zs7ket7tpeu5bnu/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.35.png?dl=0

https://www.dropbox.com/s/9hhpkddp5qr50cb/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.42.png?dl=0

https://www.dropbox.com/s/jqh3acq6dz7tams/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.47.png?dl=0

**Comment by prasoc** [comment]
ZIP the "0030700" folder for me, there are some saved replay files in there inside the "000000001"

**Comment by Ncapron** [comment]
[00030700.zip](https://github.com/citra-emu/citra/files/714914/00030700.zip)

**Comment by Ncapron** [comment]
i saved the game when i killed Ghoma so you just have to go the the room and use the blue portal and talk with the deku tree

**Comment by prasoc** [comment]
Thank you, my version is 00033500 (english european), do you have the French version?

**Comment by Ncapron** [comment]
yes i got the french one

**Comment by Ncapron** [comment]
but when you'll talk to the tree, you just have press "A" again and again and he'll ask if you want to know more ?
-> yes (oui)
->no (non)

and whatever the choice he will say ok there is a curse and it crashed

**Comment by TTFH** [comment]
There's a corrupt shader with block_size: 235678752

Vertex Shader
```
#version 330 core
layout(location = 0) in vec4 vert_position;
layout(location = 1) in vec4 vert_color;
layout(location = 2) in vec2 vert_texcoord0;
layout(location = 3) in vec2 vert_texcoord1;
layout(location = 4) in vec2 vert_texcoord2;
layout(location = 5) in float vert_texcoord0_w;
layout(location = 6) in vec4 vert_normquat;
layout(location = 7) in vec3 vert_view;

out vec4 primary_color;
out vec2 texcoord[3];
out float texcoord0_w;
out vec4 normquat;
out vec3 view;

void main() {
    primary_color = vert_color;
    texcoord[0] = vert_texcoord0;
    texcoord[1] = vert_texcoord1;
    texcoord[2] = vert_texcoord2;
    texcoord0_w = vert_texcoord0_w;
    normquat = vert_normquat;
    view = vert_view;
    gl_Position = vec4(vert_position.x, vert_position.y, -vert_position.z, vert_position.w);
}
```

Fragment Shader

```
#version 330 core
#define NUM_TEV_STAGES 6
#define NUM_LIGHTS 8

// Texture coordinate offsets and scales
#define OFFSET_256 (0.5 / 256.0)
#define SCALE_256 (255.0 / 256.0)

in vec4 primary_color;
in vec2 texcoord[3];
in float texcoord0_w;
in vec4 normquat;
in vec3 view;

in vec4 gl_FragCoord;

out vec4 color;

struct LightSrc {
    vec3 specular_0;
    vec3 specular_1;
    vec3 diffuse;
    vec3 ambient;
    vec3 position;
    float dist_atten_bias;
    float dist_atten_scale;
};

layout (std140) uniform shader_data {
    vec2 framebuffer_scale;
    int alphatest_ref;
    float depth_scale;
    float depth_offset;
    int scissor_x1;
    int scissor_y1;
    int scissor_x2;
    int scissor_y2;
    vec3 fog_color;
    vec3 lighting_global_ambient;
    LightSrc light_src[NUM_LIGHTS];
    vec4 const_color[NUM_TEV_STAGES];
    vec4 tev_combiner_buffer_color;
};

uniform sampler2D tex[3];
uniform sampler1D lut[6];
uniform usampler1D fog_lut;

// Rotate the vector v by the quaternion q
vec3 quaternion_rotate(vec4 q, vec3 v) {
    return v + 2.0 * cross(q.xyz, cross(q.xyz, v) + q.w * v);
}

void main() {
vec4 primary_fragment_color = vec4(0.0);
vec4 secondary_fragment_color = vec4(0.0);
discard; }
```

**Comment by Ncapron** [comment]
do I have to change the Vertex Shrader to the Fragment one?

if yes, Where it is? 

thank you guys

**Comment by wwylele** [comment]
@Ncapron no, that Vertex and Fragment Shader dump is for developers to locate the problem. There is nothing for you to do now

**Comment by Ncapron** [comment]
hey guys, do you have news about this?

Thank you for your time

**Comment by prasoc** [comment]
i've been unable to test your savefile, tried both USA region ROM and EU region ROM both don't have the right save ID
could you provide futher details on what version of the ROM you are using is?

**Comment by PPLToast** [comment]
The titleid 00030700 from Ncapron is for Mario Kart 7 and not OoT3D. He might be checking the wrong save folders. Also prasoc, you said your OoT3D was "european english", but the id you posted (00033500) is for the american version??

**Comment by prasoc** [comment]
I have got EUR 00033600 too, just waiting for Ncapron to provide the real save now then :)

**Comment by Ncapron** [comment]
I just have the one that I've sent :/
you can see the path with those pictures

https://www.dropbox.com/s/39s7hziampuqqel/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.15.png?dl=0

https://www.dropbox.com/s/h0zb65epl9721zf/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.21.png?dl=0

https://www.dropbox.com/s/cklma7g673gjwly/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.28.png?dl=0

https://www.dropbox.com/s/zs7ket7tpeu5bnu/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.35.png?dl=0

https://www.dropbox.com/s/9hhpkddp5qr50cb/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.42.png?dl=0

https://www.dropbox.com/s/jqh3acq6dz7tams/Capture%20d%27%C3%A9cran%202017-01-18%2020.49.47.png?dl=0

**Comment by SigmaVirus** [comment]
This must be Mac related because I've been able to continue and progress after this event on Windows 10 on a recent bleeding-edge build just fine (Ocarina of Time 3D European version) and I'm heading up to Death Mountain now in the game.

**Comment by fabikw** [comment]
It's not a mac issue. It happens to me in Win 7.

**Comment by mailwl** [comment]
my win10 kill itself with blue screan
dont know hou to fix it

**Comment by fabikw** [comment]
Couldn't upload, so here's a link:
[OoT Save Folder](https://www.dropbox.com/s/qc24yhxn29oikvb/00033500.zip?dl=0)

Just go through the dungeon (drop through hole, climb block to the left, drop through hole, enter door) to the boss room (already cleared) and get into the portal, try to listen to story of Deku Tree

**Comment by RavenHome1** [comment]
same here, windows 7 the emulator crashed

**Comment by Schplee** [comment]
This may be a hardware issue, because the most recent Nightly (e10b11a) and BE-202 do not crash after/while talking with the Deku tree. I used @fabikw's save file, tested on Windows 10.

**Comment by fabikw** [comment]
Haven't tried since BE 194 and the nightly from that day. I'll try again with the new version.

**Comment by TTFH** [comment]
You need to diseable Shader-JIT in that point; every version crash because tex[3] and lut[6] are unimplemented.

**Comment by fabikw** [comment]
I tried BE 202, disabled Shader JIT and still got the same crash.
@Schplee did you go through the dungeon to the end to have the Deku Tree cutscene? This cutscene is not triggered by going out of the dungeon through the entrance and talking with the tree.

**Comment by FernandoS27** [comment]
Shader-JIT is unrelated to that. The 3ds emulated shaders are run on the cpu. Those are the OpenGL shaders giving troubles due to a mismatch between what we are inserting into them and what they expect.

**Comment by fabikw** [comment]
Ok, so as of now, there's no way for them to work then?

**Comment by FernandoS27** [comment]
get to the moment before the crash and disable the hardware renderer.

**Comment by fabikw** [comment]
That lets me go through the cutscene (or I assume so, as I stopped it because it makes it ridiculously slow).

**Comment by FernandoS27** [comment]
yeah, you'll have to leave it that way until the cutscene fully ends.

**Comment by fabikw** [comment]
Yeah... 
Is this something that needs to be implemented or a bug that needs to be fixed?

**Comment by Schplee** [comment]
@fabikw I know how the cutscene is triggered. Yes, I went through the blue portal in the dungeon where you defeat Gohma. In both the Nightly and the Bleeding Edge, I was able to view the cutscene without a single issue. I am going to perform a couple tests and edit this post with anything I've found.

**Comment by FernandoS27** [comment]
hmn, could you get us your gpu specs using gpu-z ?

**Comment by fabikw** [comment]
[This](https://www.dropbox.com/s/7gcdg0ywp64xun1/log.txt?dl=0) is a dropbox link to my log. Too big for pastebin

My laptop has two gpus, An integrated one and a regular one. Don't know how to check which one citra is using.

![gpu](https://cloud.githubusercontent.com/assets/20501878/22956834/6395ca50-f2f2-11e6-97da-4799fd532339.gif)

![gpu2](https://cloud.githubusercontent.com/assets/20501878/22956839/6c5ceede-f2f2-11e6-8895-fb1dfcf46d27.gif)

**Comment by Schplee** [comment]
@fabikw The link to your log file says the file is no longer there. Both of your GPU's support OpenGL 4+ so I don't see an issue with compatibility. I'd still like to be able to see your log file though.

**Comment by fabikw** [comment]
I reuploaded the file, but it stops at 2 seconds. I can do another one with a different log filter later.

**Comment by FernandoS27** [comment]
hmn, could anyone else having this issue post their gpu specs? I need to check if it is AMD related.

**Comment by TTFH** [comment]
On nVidia (tested on GTX 650 and GTX 1080) the crash doesn't happen but in Intel it happen 100% of the time (at least on Celeron N2830)

**Comment by fabikw** [comment]
I got past it on an nVidia card.

**Comment by JayFoxRox** [comment]
Stop these useless lengthy discussions (or even worse: speculation) please. If anybody is able to reproduce the issue they should upload a proper trace of what happens or a good dump of the last compiled shader.

We don't care if anybody found a savegame or temporarily switched to another GPU or s/w rendering to work around the issue. We should care about fixing the issue and getting data to work with. So *please* provide an [apitrace trace](https://github.com/apitrace/apitrace) file. I've followed this issue for months and nobody seems to be able to produce a good trace / description.

**Comment by TTFH** [comment]
Here is a apitrace dump before the crash (just the last 600 lines).

[log.txt](https://github.com/citra-emu/citra/files/784025/log.txt)

**Comment by JayFoxRox** [comment]
I've tried to clean up the thread a little. @citra-emu/maintainers please shut down disucssions earlier next time and refer people to the forums. If people are not skilled enough to upload save files or debug issues they should not post in these issues.

---

Thanks a lot @TTFH , sorry again, I didn't see your earlier post (due to all the other noise).
I'm not sure where you got the idea about `tex[3]` and `lut[6]` being problematic.
The problem is pretty obvious: `shader_data` is optimized out as it's not used in the shader:

```
1327 glUseProgram(program = 180)
1328 glGetUniformLocation(program = 180, name = "tex[0]") = -1
1329 glGetUniformLocation(program = 180, name = "tex[1]") = -1
1330 glGetUniformLocation(program = 180, name = "tex[2]") = -1
1331 glGetUniformLocation(program = 180, name = "lut[0]") = -1
1332 glGetUniformLocation(program = 180, name = "lut[1]") = -1
1333 glGetUniformLocation(program = 180, name = "lut[2]") = -1
1334 glGetUniformLocation(program = 180, name = "lut[3]") = -1
1335 glGetUniformLocation(program = 180, name = "lut[4]") = -1
1336 glGetUniformLocation(program = 180, name = "lut[5]") = -1
1337 glGetUniformLocation(program = 180, name = "fog_lut") = -1
1338 glGetUniformBlockIndex(program = 180, uniformBlockName = "shader_data") = 4294967295
1339 glGetActiveUniformBlockiv(program = 180, uniformBlockIndex = 4294967295, pname = GL_UNIFORM_BLOCK_DATA_SIZE, params = 0x1c1fe0e0)
```

In line 1338 you can see that querying the `shader_data` variable fails and returns -1 (4294967295). Querying the active uniform block with index -1 does fail and return a broken size.

There are 2 steps necessary to fix this issue:

1. Be more careful when querying uniform blocks by checking if their index is even valid (so anything but GL_INVALID_INDEX). This would fix the issue and improve code quality.
2. Find out why the shader is even compiled. I can't think of a realistic situation where you'd want to make a draw call and abort writing any pixels (except for pixel counting / early-out z-tests which we don't support yet. Maybe stencil?). It looks to me like we should optimize out the entire draw-call. It's also possible that something else goes wrong and causes this bizarre rendering state (even if it doesn't crash, it still means wasted time on a draw call).

**Comment by jroweboy** [comment]
Fixed by #2581

**Comment by RavenHome1** [comment]
can confirm that it is working now
